<center> <h1>The tutorial to use ChipSeq</h1> </center>


##Set up environment

Install the following softares firstly:

1. ngsplt(https://github.com/shenlab-sinai/ngsplot)

2. MACS(https://github.com/taoliu/MACS)

3. ChipSeq(https://github.com/aiminy/ChipSeq)

```{r eval=FALSE}

#To download

git clone https://github.com/aiminy/ChipSeq

#To build 

R CMD build --resave-data ChipSeq

#To Install
R CMD INSTALL ChipSeq_0.0.1.tar.gz -l ~/R/lib64/R/library/
```

##Peak Call and Annotation
```{r eval=FALSE}

input.file.dir="/projects/scratch/bbc/Project/Danny_chip/Alignment/BWA/"
input.file.pattern="*.bam"
index.file=9
output.file.dir="/scratch/projects/bbc/aiminy_project/
genome="Hs"
 
library(ChipSeq)
PeakCallAndAnnotation(input.file.dir,input.file.pattern,index.file,output.file.dir,genome)
```

##Bam files sorting, index and Visualization

+ Run ChipSeq to perform one-setp analysis for ChipSeq related data 

```{r eval=FALSE}
input.file.dir="/projects/scratch/bbc/Project/Danny_chip/Alignment/BWA/"
input.file.pattern="*.bam"
index.file=9
output.file.dir="/scratch/projects/bbc/aiminy_project/
genome="Hs"
 
BamFileSortIndexVisualization(input.file.dir,input.file.pattern,index.file,output.file.dir,genome)(input.file.dir,input.file.pattern,index.file,output.file.dir,genome)
 
```

## Perform the above steps in one streamlined procedure

If you run based on bash
```{bash eval=FALSE}
sh ~/Code/BashRunMACS1-4-2_4_Danny_chip_seq3.sh /projects/scratch/bbc/Project/Danny_chip/Alignment/BWA/ 9 ".bam" /scratch/projects/bbc/aiminy_project/ "Hs"
```

If you run based on R
```{r eval=FALSE}

Rscript R_code/Set_up_R.R
Rscript R_code/Run_Chip_Seq.R $DIR $INDEX $pattern $results_dir $out 

```
